### PR TITLE
EAGLE3.1 Support

### PR DESCRIPTION
**Pull request by bluecoffee8**
## Motivation

EAGLE3.1 support, based on [https://github.com/lightseekorg/TorchSpec/pull/97](url) which was added to torchspec. 

Validation: trained Qwen3-30B-A3B-Instruct-2507 draft model (EAGLE3 vs EAGLE3.1) on a single epoch of sharegpt dataset, based on [https://github.com/sgl-project/SpecForge/blob/main/examples/run_qwen3_30b_a3b_eagle3_online.sh](url). 

## Modifications

Add eagle3.1 features (draft model output norm, fc norm on target hidden states). 
Added example eagle3.1 config for qwen3-30b-a3b model, and script. 

## Benchmark & Profiling

Ran benchmarks for each. 

**Server launch (1 x H100):** 
```
SGLANG_ENABLE_SPEC_V2=1 && SGLANG_ALLOW_OVERWRITE_LONGER_CONTEXT_LEN=1 && python -m sglang.launch_server --host 0.0.0.0 --port 8001 --model-path Qwen/Qwen3-30B-A3B-Instruct-2507 --speculative-algorithm EAGLE3 --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 --speculative-draft-model-path /path/to/draft_model
```

**Client launch:**
```
python3 -m sglang.bench_serving --backend sglang --host 0.0.0.0 --port 8001 --dataset-name random-ids --warmup-requests 80 --num-prompts 128 --random-input-len 1024 --random-output-len 256 --random-range-ratio 1.0 --request-rate 1.0 
```

**Results**: 
EAGLE3

img_v3_02125_ff5f39c2-f740-4b5d-8083-a4ab77b98a6h

EAGLE3.1

img_v3_02125_873e5d0d-054c-46d2-a52b-d89d67dc73ch

**Comparison**: 

Acc length 1.68 => 2.30, +36%
p50 e2e, 1550.76 => 823.57, -47%
p50 tpot, 5.75 => 2.94, -49%

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

File: specforge/core/eagle3.py [hunk 1]
```diff
@@ -538,6 +542,10 @@ def forward(
             # update hidden states for next step
             hidden_states = hidden_states_out
 
+            # Apply output norm for EAGLE 3.1 post-norm architecture
+            if self.draft_model.norm_output:
+                hidden_states = self.draft_model.norm(hidden_states)
+
             # Step 5.4: get logits
             logits = self.draft_model.compute_logits(hidden_states)
```
Comment (Dogacel) [inline specforge/core/eagle3.py]: `compute_logits` already applies norm to the `hidden_states`. We should gate it if the hidden states is already normed.

File: specforge/modeling/draft/llama3_eagle.py [hunk 2]
```diff
@@ -1406,8 +1416,12 @@ def embed_input_ids(self, input_ids: torch.Tensor) -> torch.Tensor:
         return self.embed_tokens(input_ids)
 
     def project_hidden_states(self, hidden_states: torch.Tensor) -> torch.Tensor:
-        # eagle 3 requires hidden states from 3 layers
-        assert hidden_states.size(-1) == self.config.hidden_size * 3
```
Comment (Dogacel) [inline specforge/modeling/draft/llama3_eagle.py:1728]: Let's keep assertion and assert to `self.num_aux_hidden_states`.

File: specforge/core/eagle3.py [hunk 3]
```diff
@@ -249,6 +249,10 @@ def forward(
             # update hidden states for next step
             hidden_states = hidden_states_out
 
+            # Apply output norm for EAGLE 3.1 post-norm architecture
+            if self.draft_model.norm_output:
+                hidden_states = self.draft_model.norm(hidden_states)
```
Comment (Dogacel) [inline specforge/core/eagle3.py]: Similarly,

I think it makes sense to calculate using `hidden_states, hidden_states_for_logits = get_hidden_states(...)` as a pair, where the right one is normed and left is not by default. The method can return both normed based on `norm_output`.

Comment (bluecoffee8) [reply specforge/core/eagle3.py]: @Dogacel i just pushed a commit that should address your comments. could you plz take a look thanks!

**Review by Dogacel — APPROVED** [review]
LGTM 👍

**Comment by xiangqianzsh** [comment]
@bluecoffee8  Hi, I have trained an eagle3.1 model using this patch. Which branch or commit id should I use when testing performance in sglang?

**Comment by bluecoffee8** [comment]
@xiangqianzsh I think you should be able to use any, the draft model trained here should just run with EAGLE-3 spec decoding option in sglang. 

@Dogacel could you confirm there should be no changes to the inference engine side of things required?

**Comment by Dogacel** [comment]
> @xiangqianzsh I think you should be able to use any, the draft model trained here should just run with EAGLE-3 spec decoding option in sglang. 
> 
> 
> 
> @Dogacel could you confirm there should be no changes to the inference engine side of things required? 

Yeah it should work out of the box.

**Comment by jiapingW** [comment]
Hi, thank you for implementing eagle3.1. I'd like to know why your bench is using random_ids. Comparing accept lengths with random data as input doesn't make much sense, does it? @bluecoffee8

**Comment by bluecoffee8** [comment]
@jiapingW you can try with other benchmark method like sharegpt dataset.

**Comment by jiapingW** [comment]
> @jiapingW you can try with other benchmark method like sharegpt dataset.

Can you opensource your eagle3.1 draft model? We can test it.

**Comment by bluecoffee8** [comment]
@jiapingW I don't have option to open source the model, but let me try to benchmark it when I have the bandwidth. If anybody else has the GPU resource they could try to train and benchmark in the meantime as well.

**Comment by jiapingW** [comment]
Hi, we've been refactoring the code recently, and it's mostly finished. Could you rebase your PR to the current `main` branch? Thanks.

**Comment by bluecoffee8** [comment]
@jiapingW just rebased.

**Review by jiapingW — APPROVED** [review]
